File: app/src/main/java/com/demo/room/dao/BaseDao.java
package com.demo.room.dao;

import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Update;
import io.reactivex.Completable;

/**
 * Created by guoxiaodong on 2020/3/21 10:55
 */
public interface BaseDao<T> {
    /**
     * 一次插入单条数据或多条
     * OnConflictStrategy.REPLACE:替换旧数据
     * OnConflictStrategy.ROLLBACK:回滚事务
     * OnConflictStrategy.ABORT:终止事务崩溃
     * OnConflictStrategy.FAIL:事务失败
     * OnConflictStrategy.IGNORE:忽略冲突
     */
    @Insert(onConflict = OnConflictStrategy.ABORT)
    Completable insert(T... ts);

    /**
     * 一次删除单条数据或多条
     */
    @Delete
    Completable delete(T... ts);

    /**
     * 一次更新单条数据或多条
     */
    @Update
    Completable update(T... ts);
}

File: app/src/main/java/com/gxd/demo/room/dao/BaseDao.kt
package com.gxd.demo.room.dao

import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Update
import io.reactivex.Single

interface BaseDao<T> {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insert(obj: T): Long

    @Insert
    fun insertRx(obj: T): Single<Long>

    @Insert
    fun insertRx(objList: List<T>): Single<List<Long>>

    @Insert
    fun insert(objList: List<T>)

    /**
     * 返回成功行数
     */
    @Update
    fun update(obj: T): Single<Int>

    @Delete
    fun delete(obj: T): Single<Int>
}